### ISSUE TITLE
Khan Academy (English) Import Metrics

### ISSUE THREAD

**Issue by tim-moody** [learningequality/kolibri#6684]
## Observed behavior

Download is quite slow and gets slower over time

2% - 10 min
10% - 80 min
30% - 6 3/4 hours
60% - 17 3/4 hours

still downloading

Anecdotally cpu has steadily increased from 60 or 70% to 90 - 100% (I take it total could be 400% with 4 cores) as reported by htop. Memory usage has grown from around 580M to 630M. Both are variable.

### Steps to reproduce

Navigate to Import and select all Khan Academy English resources (12638 resources (39 GB))

### Context

RPi 4 with 2 G RAM and 128G sdcard
64 bit Ubuntu 19.10.1
wired internet connection on fast connection (1Gbs)
Kolibri 0.13.1

**Comment by tim-moody** [comment]
At 24 hours just short of 75% complete.

**Comment by tim-moody** [comment]
From 0 to 10% took 80 minutes.
From 70 to 80% took 250.

**Comment by tim-moody** [comment]
It completed in just short of 36 hours for an average speed of 18 MB per minute.

By way of comparison, the Spanish Khan Academy (17G) subsequently completed in 90 minutes on the same machine for an average speed of 190 MB per minute.

**Comment by holta** [comment]
> It completed in just short of 36 hours for an average speed of 18 MB per minute.

Depressing, but an essential data point: that's 2.4 mbit/sec, similar to late 1990s Internet speeds.

Is anybody at Learning Equality able to explain a bit more what download protocol and URL's are causing this, for investigation?

**Comment by indirectlylit** [comment]
Thanks for the info. It's super curious that KA (ES) completed so much quicker than KA (EN).

We did recently have an import performance regression (https://github.com/learningequality/kolibri/issues/6602), and have a fix for it coming out in 0.13.2, but the behavior sounds different from what you're describing.

That said, perhaps you'd be able to test our 0.13.2 beta to help ascertain if any performance characteristics have changed?

https://github.com/learningequality/kolibri/releases/tag/v0.13.2-beta1

thank you!

**Comment by indirectlylit** [comment]
Another issue related to memory usage: https://github.com/learningequality/kolibri/issues/2879

**Comment by kollivier** [comment]
How often during the KA import process did you see the high CPU? Download of content should not be CPU intensive, so it suggests that there is some post-processing of the downloaded data that is getting exponentially slower over time. Does the high CPU happen right from the start, or creep upwards over time, or is it really variable? (Sorry, know the process has already finished, so whatever you can remember would be helpful!)

**Comment by benjaoming** [comment]
@tim-moody do you have any logging output from the time of the import? I'm asking because I'm currently seeing a disruption between Cloudflare and our origin server, resulting in frequent 522 server errors which will show up in logs (`~/.kolibri/logs/kolibri.txt` or `~/.kolibri/logs/archive`).

Try searching logs for this:

> An error occurred during content import

**Comment by tim-moody** [comment]
@kollivier I wrote down subjective notes on the cpu% which had a min, a max, and a cluster. The cluster over time, gradually and slowly, increased from 80 to 90 to 100% (this is one of 4 cores). For the Spanish download the cluster was more like 40 to 50% and didn't seem to increase.

**Comment by tim-moody** [comment]
@benjaoming here are the logs
https://paste.ubuntu.com/p/QTCWxG9h6N/
https://paste.ubuntu.com/p/K8Q9kzRkMH/
https://paste.ubuntu.com/p/8vRCp4SBbh/
https://paste.ubuntu.com/p/9N7ChythxT/
https://paste.ubuntu.com/p/ytxGkKpcqM/

**Comment by benjaoming** [comment]
Thanks @tim-moody - It doesn't seem like we have logged any errors, but I would rule out the "522 Server Error" then. I cannot figure out the timing of events from the logs, but channel id `1ceff53605e55bef987d88e0908658c5` is Khan Academy English -- however doesn't seem to be any logged event for finishing the download.

```
INFO 2020-03-21 19:46:34,430 importchannel Downloading data for channel id 1ceff53605e55bef987d88e0908658c5
INFO 2020-03-21 19:46:42,412 channel_import Importing ContentTag data
INFO 2020-03-21 19:46:42,414 channel_import Importing ContentNode_has_prerequisite data
INFO 2020-03-21 19:46:42,415 channel_import Importing ContentNode_related data
INFO 2020-03-21 19:46:42,416 channel_import Importing ContentNode_tags data
INFO 2020-03-21 19:46:42,417 channel_import Importing ContentNode data
INFO 2020-03-21 19:46:45,363 channel_import Importing Language data
INFO 2020-03-21 19:46:45,365 channel_import Importing File data
INFO 2020-03-21 19:47:09,057 channel_import Importing LocalFile data
INFO 2020-03-21 19:47:09,980 channel_import Importing AssessmentMetaData data
INFO 2020-03-21 19:47:10,239 channel_import Importing ChannelMetadata data
INFO 2020-03-21 19:47:23,493 channel_import Channel 1ceff53605e55bef987d88e0908658c5 successfully imported into the database
```

**Comment by benjaoming** [comment]
Ping @mrpau-richard - you are reproducing a database issue that happens after importing 20 gigs of Khan Academy, can you notice during the next test if it also slows down during import as described here?

**Comment by benjaoming** [comment]
Ref: https://github.com/learningequality/kolibri/issues/6703

**Comment by tim-moody** [comment]
Starting a new run on Intel NUC running Ubuntu 18.04.4 LTS. Kolibri 0.13.1
Pace is slower than before. at 12 minutes still not to 2%.
CPU < 20%

**Comment by tim-moody** [comment]
make that 20 min and still at 1%

**Comment by benjaoming** [comment]
@tim-moody -- thanks for this, any other kinds of profiling are helpful to know the symptoms. As you can see from https://github.com/learningequality/kolibri/issues/6715, it is quite likely that the slowness isn't caused by downloading but rather by the import process itself.

* ~Have you tried installing something to measure I/O on network? Is the bandwidth jumpy or constant? `iftop` does the trick, there are also [others](https://askubuntu.com/questions/257263/how-to-display-network-traffic-in-the-terminal)~
* ~Is the media / SD card full and causing slow I/O?~

~(this is just to rule out some of the scenarios that seem less likely at present)~

Edit: No reason to test this, I think the symptoms are well described and debugging is already taking place.

**Comment by tim-moody** [comment]
85% at 24 hours

**Comment by kollivier** [comment]
@tim-moody We have just merged a PR (#7521) that should provide significant improvements to channel download speeds for channels like KA. These fixes will be in the upcoming 0.14.4 release.

**Comment by rtibbles** [comment]
This should be fixed by #7521 - please advise if this is not the case.
